Add use-forwarded-headers configmap option.

Dirbaio opened the pull request:
This PR adds a configmap option `use-forwarded-headers` to enable/disable trust of incoming `X-Forwarded-*` headers. 

Setting it to `true` corresponds to the current ingress behavior.

It is set to `false` by default. **this is a breaking change**, but IMO it's the right thing to do if we follow the "secure by default" principle:

- If the default is `true`, it might lead to vulnerable installations: If it's set to true in a scenario when it shouldn't (direct port forwarding, L3 load balancers) the config is vulnerable to source IP spoofing. However, since everything appears to work correctly, the user probably won't notice, and will end up vulnerable in production.
- If the default is `false`, it'll just cause incorrect behavior if it's wrong. The user is likely to notice he's getting incorrect source IPs and will set it to `true`. At no point the user is vulnerable to spoofing.

Fixes #1815 #1309 #1668
Rebase/update of #1851

antoineco commented:
This is indeed a breaking change, and not the behaviour most people will expect (why would my web proxy decide for me whether a header is secure or not?). If this gets approved we should first leave this to `true` and log a verbose message notifying users this default will change in future versions.

Dirbaio commented:
> why would my web proxy decide for me whether a header is secure or not?

It has to decide something if the user doesn't. IMO it's better to go with the most conservative option, which is assuming XFF headers aren't secure and not trusting them. 

If doing breaking changes is not OK then I guess logging a warning is the best we can do, though..

I'll add unit tests and docs soon.

antoineco commented:
I'm perfectly with you @Dirbaio, just trying to think about a safe path so that users can be warned im a timely manner before we do change the default (not everybody reads release notes carefully :wink:).
The change to `false` by default can be done in 2 versions instead of 1.

Dirbaio commented:
- Changed setting to be true by default.
- Added e2e test that checks for both behaviors
- Added docs.

I haven't added the warning when the option is unset though because I'm unsure where to add it :) Help would be appreciated (or we can just not add it?)

Dirbaio commented:
CI failed due to something unrelated (I think). How do I retest? 

also ping @antoineco PTAL

antoineco commented:
I restarted it.

antoineco commented on docs/user-guide/nginx-configuration/configmap.md:
```diff
@@ -596,6 +597,12 @@ _References:_
 
+## use-forwarded-headers
+
+If true, nginx passes the incoming `X-Forwarded-*` headers to upstreams. Use this option when nginx is behind another L7 proxy / load balancer that is setting these headers.
```
(nitpick) NGINX is commonly written in capital letters.

Dirbaio replied:
Done

antoineco commented on rootfs/etc/nginx/template/nginx.tmpl:
```diff
@@ -222,7 +226,9 @@ http {
     }
 
-    map {{ buildForwardedFor $cfg.ForwardedForHeader }} $the_real_ip {
+    # The following is a sneaky way to do "set $the_real_ip $remote_addr"
```
I'd rather use a Go template comment to avoid writing this message to the config file.

Dirbaio replied:
IMO this could be useful to someone `cat`ing the rendered config from a running container, to help understand what's going on. 

The rendered template already contains other similar comments, and this one is rendered just once so it's not like it'd cause much bloat.

antoineco replied:
You're right, no strong opinion on that one.

antoineco commented on test/e2e/settings/forwarded_headers.go:
```diff
@@ -0,0 +1,123 @@
+		Expect(err).NotTo(HaveOccurred())
+		defer conn.Close()
+
+		conn.Write([]byte("GET / HTTP/1.1\r\nHost: forwarded-headers\r\nX-Forwarded-Port: 1234\r\nX-Forwarded-Proto: myproto\r\nX-Forwarded-For: 1.2.3.4\r\nX-Forwarded-Host: myhost\r\n\r\n"))
```
Using the `gorequest` package may make this easier to read.

Dirbaio replied:
Done

antoineco commented:
Overall LGTM, however I have a feeling I'm not experienced enough with the NGINX `map` directive to express a final statement regarding all this map-fu :)

antoineco commented:
@aledbf I need your opinion on 2 things:

1. Is this reasonable to you to change the default in a future version?
If yes @Dirbaio please add the warning similarly to https://github.com/kubernetes/ingress-nginx/blob/master/cmd/nginx/flags.go#L213

2. The `map` logic added to the template.

Dirbaio commented:
@antoineco the code you linked is for cmdline flags, but this is a configmap flag, which is pulled dynamically from the Kubernetes API.

antoineco commented:
Yeah stupid me, ignore that comment.

The next best place would be inside the `OnUpdate()` function then.

Dirbaio commented:
ping @aledbf . Does the PR look OK? Should I add the future change warning?

aledbf commented:
@Dirbaio I will take a look at this PR after we release 0.16. Apologies for the delay

albertvaka commented:
Seems this got ignored again like it happened with #1851

Not great for a security issue :/

antoineco commented:
/assign aledbf 
https://github.com/kubernetes/ingress-nginx/pull/2616#issuecomment-399277140

aledbf commented:
/lgtm
/approve

aledbf commented:
@Dirbaio again, apologies for the delay with the review and merge of this PR

aledbf commented:
@Dirbaio thanks!

aledbf commented:
/hold

ElvinEfendi commented:
Thanks for the PR @Dirbaio!

I wonder if we can get rid of `ComputeFullForwardedFor` and `X-Original-Forwarded-For` at all. There's a case where we do
```
{{ $proxySetHeader }} X-Forwarded-For        $the_real_ip;
```
Why would a reverse proxy do this? This seems like a hacky way of satisfying some lazy upstream so that they can get the correct client IP.

Dirbaio commented on rootfs/etc/nginx/template/nginx.tmpl:
```diff
@@ -1047,7 +1069,7 @@ stream {
             {{ $proxySetHeader }} X-Request-ID           $req_id;
             {{ $proxySetHeader }} X-Real-IP              $the_real_ip;
-            {{ if $all.Cfg.ComputeFullForwardedFor }}
+            {{ if and $all.Cfg.UseForwardedHeaders $all.Cfg.ComputeFullForwardedFor }}
```
`$full_x_forwarded_for` is the same as `$proxy_add_x_forwarded_for` but done "by hand" in a map above.

Maybe this is so it appends the proxy protocol addr instead of `$remote_addr` if using proxy protocol, and nginx's `$proxy_add_x_forwarded_for` doesn't?

ElvinEfendi replied:
yeah that makes sense @Dirbaio 

Why do you require `$all.Cfg.UseForwardedHeaders` to be `true` here though?

Dirbaio replied:
`ComputeFullForwardedFor` has no effect when `UseForwardedHeaders = false`. 

There are 3 cases to consider:

- `UseForwardedHeaders = false`: incoming X-Forwarded-For value is discarded and replaced with the source IP nginx-ingress sees.
- `UseForwardedHeaders = true, ComputeFullForwardedFor = false`: incoming X-Forwarded-For value is passed through if present, otherwise it's filled with the source IP nginx-ingress sees.
- `UseForwardedHeaders = true, ComputeFullForwardedFor = true`: incoming X-Forwarded-For value is passed through, appended with the source IP nginx-ingress sees.

For case 3 we use `$full_x_forwarded_for`, which has been computed higher in the template. 
For cases 1, 2 we use `$the_real_ip`, which higher in the template has been set to be the appropriate source IP depending on `UseForwardedHeaders`

schweikert commented:
We noticed by chance that the default behavior of ingress-nginx is to trust the X-Forwarded-For header, which caused a security vulnerability for our setup. Is there an issue tracking the change of the default to false for this? I think that it would be much preferable to follow a "secure by default" approach. How many users of ingress-nginx are affected without knowing?

ElvinEfendi commented:
@schweikert unfortunately that's the case. I think default value for`proxy-real-ip-cidr` should be ` "0.0.0.0/32"`, so that it trusts no client.

https://github.com/kubernetes/ingress-nginx/pull/3333 aims to do so.
